odl/solvers/functional/default_functionals.py

```diff
@@ -0,0 +1,731 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Default functionals defined on any (reasonable) space."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+from odl.solvers.functional.functional import Functional
+from odl.operator.operator import Operator
+from odl.solvers.advanced.proximal_operators import (
+    proximal_l1, proximal_cconj_l1, proximal_l2, proximal_cconj_l2,
+    proximal_l2_squared)
+
+from odl.operator.default_ops import (ZeroOperator, IdentityOperator)
+
+
+__all__ = ('L1Norm', 'L2Norm', 'L2NormSquare', 'ZeroFunctional',
+           'ConstantFunctional')
+
+
+# TODO: Implement some of the missing gradients
+
+class L1Norm(Functional):
+
+    """The functional corresponding to L1-norm.
+
+    The L1-norm, ``||x||_1``,  is defined as the integral/sum of ``|x|``.
+
+    References
+    ----------
+    Wikipedia article on `Norms
+        <https://en.wikipedia.org/wiki/Norm_(mathematics)>`_.
+    """
+
+    def __init__(self, domain):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain : `LinearSpace`
```

Comment:
The current implementation cannot handle `ProductSpace`. But something like
```
x.ufunc.absolute().inner(self.domain.one())
```
lets it pass. However, the following example
```
>>> import odl
>>> space = odl.ProductSpace(odl.uniform_discr(0,1,2), odl.uniform_discr(0,1,3))
>>> tmp = odl.solvers.L1Norm(space)
>>> element = -1*space.one()
>>> tmp(element)
2.0
```
shows that it might not really work as intended... Is `2.0` or `1.0` expected? Change to this formulation?